[Sema] Substitute parameter packs when deduced from function arguments

spaits opened the pull request:
This pull request would solve https://github.com/llvm/llvm-project/issues/78449 .
There is also a discussion about this on stackoverflow: https://stackoverflow.com/questions/77832658/stdtype-identity-to-support-several-variadic-argument-lists .

The following program is well formed:
```cpp
#include <type_traits>

template <typename... T>
struct args_tag
{
    using type = std::common_type_t<T...>;
};

template <typename... T>
void bar(args_tag<T...>, std::type_identity_t<T>..., int, std::type_identity_t<T>...) {}

// example
int main() {
    bar(args_tag<int, int>{}, 4, 8, 15, 16, 23);
}
```

but Clang rejects it, while GCC and MSVC doesn't. The reason for this is that, in `Sema::DeduceTemplateArguments` we are not prepared for this case. 

# Substitution/deduction of parameter packs
The logic that handles substitution when we have explicit template arguments (`SubstituteExplicitTemplateArguments`) does not work here, since the types of the pack are not pushed to `ParamTypes` before the loop starts that does the deduction.
The other "candidate" that may could have handle this case would be the loop that does the deduction for trailing packs, but we are not dealing with trailing packs here.

# Solution proposed in this PR
The solution proposed in this PR works similar to the trailing pack deduction. The main difference here is the end of the deduction cycle.
When a non-trailing template pack argument is found, whose type is not explicitly specified and the next type is not a pack type, the length of the previously deduced pack is retrieved (let that length be `s`). After that the next `s` arguments are processed in the same way as in the case of non trailing packs.

# Another possible solution
There is another possible approach that would be less efficient. In the loop when we get to an element of `ParamTypes` that is a pack and could be substituted because the type is deduced from a previous argument, then `s` number of arg types would be inserted before the current element of `ParamTypes` type. Then we would "cancel" the processing of the current element, first process the previously inserted elements and the after that re-process the current element.
Basically we would do what `SubstituteExplicitTemplateArguments` does but during deduction.

# Adjusted test cases
In `clang/test/CXX/temp/temp.fct.spec/temp.deduct/temp.deduct.call/p1-0x.cpp` there is a test case named `test_pack_not_at_end` that should work, but still does not. This test case is relevant because the note for the error message has changed.
This is what the test case looks like currently:
```cpp
template<typename ...Types>
void pack_not_at_end(tuple<Types...>, Types... values, int); // expected-note {{<int *, double *> vs. <int, int>}}

void test_pack_not_at_end(tuple<int*, double*> t2) {
  pack_not_at_end(t2, 0, 0, 0); // expected-error {{no match}}
  // FIXME: Should the "original argument type must match deduced parameter
  // type" rule apply here?
  pack_not_at_end<int*, double*>(t2, 0, 0, 0); // ok
}

```

The previous note said (before my changes):
```
deduced conflicting types for parameter 'Types' (<int *, double *> vs. <>)
````
The current note says (after my changesand also clang 14 would say this if the pack was not trailing):  
```
deduced conflicting types for parameter 'Types' (<int *, double *> vs. <int, int>)
```
GCC says: 
```
error: no matching function for call to ‘pack_not_at_end(std::tuple<int*, double*>&, int, int, int)’
   70 |     pack_not_at_end(t2, 0, 0, 9); // expected-error {{no match}}
````

spaits committed:
[Sema]Substitue template parameter packs when deduced from function parameter

spaits committed:
Format changes

spaits committed:
Adjust test case and format

spaits committed:
Add tests

cor3ntin commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -4371,6 +4397,41 @@ Sema::TemplateDeductionResult Sema::DeduceTemplateArguments(
+        // the already deduced parameter.
+        SmallVector<UnexpandedParameterPack, 2> Unexpanded;
+        collectUnexpandedParameterPacks(ParamPattern, Unexpanded);
+        if (Unexpanded.size() == 0)
+          continue;
```
can `unexpanded` ever be empty? This should be an assert

spaits replied:
I think it could be. In that case that every template packs that are needed by the dependent name are all expanded. For example this could happen when we use explicitly specified template arguments.

spaits replied:
I did some optimization and this part of the code is not needed anymore.

cor3ntin commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -4371,6 +4397,41 @@ Sema::TemplateDeductionResult Sema::DeduceTemplateArguments(
+        // replaced with the corresponding deduced or default argument values
+        //
+        // If we have a trailing parameter pack, that has been deduced
+        // perviously we substitute the pack here in a similar fashion as seen
```
```suggestion
        // previously we substitute the pack here in a similar fashion as seen
```

cor3ntin replied:
```suggestion
        // perviously we substitute the pack here in a similar fashion as
```

cor3ntin commented on clang/test/SemaTemplate/deduction.cpp:
```diff
@@ -431,6 +442,17 @@ namespace deduction_after_explicit_pack {
   }
 
+  template <typename... T>
+  void bar(args_tag<T...>, type_identity_t<T>..., int mid, type_identity_t<T>...) {}
```
What happens without the mid parameter ?

spaits replied:
It could work. It would cost only condition being removed. I deliberately put in the condition that disables this. Should I enable it?

I think enabling it would be standard compliant, but I played it safe and stuck to the example seen in the issue.

cor3ntin replied:
Yes, i think we should support that, and add a test for it )as long as the standard says it should work, which i think it does

spaits replied:
I added a test for this.

cor3ntin replied:
I'd like to see a test for ```cpp
template <typename... Y, typename... T>
void foo2(args_tag<Y...>, args_tag<T...>, type_identity_t<T>..., type_identity_t<T>...) {}
```

spaits replied:
Added the test. It also works.

cor3ntin commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -858,6 +859,30 @@ class PackDeductionScope {
       Info.PendingDeducedPacks[Pack.Index] = Pack.Outer;
   }
 
+  std::optional<unsigned> getSavedPackSize(unsigned Index,
+                                           TemplateArgument Pattern) const {
+
+    SmallVector<UnexpandedParameterPack, 2> Unexpanded;
+    S.collectUnexpandedParameterPacks(Pattern, Unexpanded);
+    if (Unexpanded.size() == 0 ||
+        Packs[0].Saved.getKind() != clang::TemplateArgument::Pack)
+      return {};
+    unsigned PackSize = Packs[0].Saved.pack_size();
+
+    if (std::all_of(Packs.begin() + 1, Packs.end(), [&PackSize](auto P) {
+          return P.Saved.getKind() == TemplateArgument::Pack &&
+                 P.Saved.pack_size() == PackSize;
+        }))
+      return PackSize;
+    return {};
+  }
+
+  /// Determine whether this pack has already been deduced from a previous
+  /// argument.
+  bool isDeducedFromEarlierParameter() const {
+    return DeducedFromEarlierParameter;
+  }
+
```
Can we save the size in PackDeductionScope instead of recomputeing it?
It should be ill-formed to deduce different sizes

spaits replied:
This code at the this state(the state in which you wrote this comment) is more complicated then it should be. I was able to make it more simple.

I gave a new name to this function, that describes better what this function does.

We are using the `Saved` packs here that are only modified in the constructors. The `Saved` packs are the previously deduced packs.

cor3ntin reviewed (commented):
Thanks for working on this!
It looks like a good direction.

I left a few comments

Can you add a changelog entry? Thanks

spaits committed:
Fix typo

Co-authored-by: cor3ntin <corentinjabot@gmail.com>

spaits committed:
Fix typo #2

spaits committed:
Substitue pack even when no separator between packs

spaits committed:
Use assertion where we know that pack is not empty

spaits committed:
Simplify logic for pack size computing

spaits committed:
Add a test case for deduction without separator

spaits commented:
Thank you @cor3ntin for reviewing my PR.

I addressed all of your comments.

- I fixed the types.
- I added support for deduction of non-trailing packs when there is no separator type.
- Simplified the code. This simplification has affect on the code you asked about.

Could you please review my recent changes?

erichkeane commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -858,6 +859,27 @@ class PackDeductionScope {
+      return {};
+    unsigned PackSize = Packs[0].Saved.pack_size();
+
+    if (std::all_of(Packs.begin() + 1, Packs.end(), [&PackSize](auto P) {
```
```suggestion
    if (std::all_of(Packs.begin() + 1, Packs.end(), [&PackSize](const auto &P) {
```

erichkeane reviewed (commented):
I don't see the release note, and there is a typo in the PR entry that should be fixed.

Else just 1 nit.

spaits commented:
> I don't see the release note, and there is a typo in the PR entry that should be fixed.
> 
> Else just 1 nit.

Oh sorry I forgot about the release note. I will update that.

spaits committed:
Add const to lambda parameter

Co-authored-by: Erich Keane <ekeane@nvidia.com>

spaits committed:
Update release notes

spaits commented:
@cor3ntin @erichkeane
Thank you for checking the updates on my PR.
I fixed the typo and added the release notes.

spaits committed:
Fix formatting od release notes

erichkeane commented:
> @cor3ntin @erichkeane Thank you for checking the updates on my PR. I fixed the typo and added the release notes. Also rebased the branch.

Looks like the release notes change failed CI.

spaits commented:
> > @cor3ntin @erichkeane Thank you for checking the updates on my PR. I fixed the typo and added the release notes. Also rebased the branch.
> 
> Looks like the release notes change failed CI.

Not my changes. This was commited 3 hours ago: https://github.com/spaits/llvm-project/commit/c92ad411f2f94d8521cd18abcb37285f9a390ecb .
In this commit the author has forgot to close a `` .

spaits commented:
It looks like libc++ test suites are failing. Looking at the logs from the CI I can not really figure out what is the exact reason.
I could only deduce that the failing test case is `libcxx/gdb/gdb_pretty_printer_test.sh.cpp`. This case seems to be related to `libcxx/selftest/convenience_substitutions/build_run.sh.cpp`. I will try to figure out how to set up my environment to run these tests.

spaits commented:
I did some more digging. I saw that the test in the CI fails with permission errors.

```
| No symbol table is loaded.  Use the "file" command.
| warning: opening /proc/PID/mem file for lwp 4103309.4103309 failed: Permission denied (13)
| Traceback (most recent call last):
|   File "/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h5ngp-1/llvm-project/clang-ci/libcxx/test/libcxx/gdb/gdb_pretty_printer_test.py", line 123, in <module>
|     gdb.execute("run")
| gdb.error: Warning:
| Cannot insert breakpoint 1.
| Cannot access memory at address 0x55555555637c
```

I thought that maybe the problem is that, the program to be debugged by `dbg` does not compile with the compiler that includes my changes. I decided to just compile the test file with my modified compiler. I copied the test file. Removed one include that is only needed for manual debugging and recreated the command used in the CI as much as I could. The command looks like this now:
```
/home/spaits/repo/llvm-project/build/bin/clang++ -pthread --target=x86_64-unknown-linux-gnu /home/spaits/cpp/deb.cpp -o /home/spaits/cpp/deb.out -std=c++26 -Werror -Wall -Wctad-maybe-unsupported -Wextra -Wshadow -Wundef -Wunused-template -Wno-unused-command-line-argument -Wno-attributes -Wno-pessimizing-move -Wno-noexcept-type -Wno-atomic-alignment -Wno-reserved-module-identifier -Wdeprecated-copy -Wdeprecated-copy-dtor -Wno-user-defined-literals -Wno-tautological-compare -Wsign-compare -Wunused-variable -Wunused-parameter -Wunreachable-code -Wno-unused-local-typedef -Wno-local-type-template-args -Wno-c++11-extensions -Wno-unknown-pragmas -Wno-pass-failed -Wno-mismatched-new-delete -Wno-redundant-move -Wno-self-move -D_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER -D_LIBCPP_ENABLE_EXPERIMENTAL -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_NONE -Werror=thread-safety -Wuser-defined-warnings -g   -lc++ -latomic
0 spaits@work-laptop:~/cpp$ 
```
Because of the GCC 13+ prereq. I did not build libc++ for my self, but used the one available on my machine.
This way the file has compiled successfully.

I will still try to run the tests myself.

Until that if you have any idea what should I, do what wen possibly wrong pleas share that with me.

Thanks.

spaits committed:
Fix grammar mistake

Co-authored-by: cor3ntin <corentinjabot@gmail.com>

erichkeane commented:
> I did some more digging. I saw that the test in the CI fails with permission errors.
> 
> ```
> | No symbol table is loaded.  Use the "file" command.
> | warning: opening /proc/PID/mem file for lwp 4103309.4103309 failed: Permission denied (13)
> | Traceback (most recent call last):
> |   File "/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h5ngp-1/llvm-project/clang-ci/libcxx/test/libcxx/gdb/gdb_pretty_printer_test.py", line 123, in <module>
> |     gdb.execute("run")
> | gdb.error: Warning:
> | Cannot insert breakpoint 1.
> | Cannot access memory at address 0x55555555637c
> ```
> 
> I thought that maybe the problem is that, the program to be debugged by `dbg` does not compile with the compiler that includes my changes. I decided to just compile the test file with my modified compiler. I copied the test file. Removed one include that is only needed for manual debugging and recreated the command used in the CI as much as I could. The command looks like this now:
> 
> ```
> /home/spaits/repo/llvm-project/build/bin/clang++ -pthread --target=x86_64-unknown-linux-gnu /home/spaits/cpp/deb.cpp -o /home/spaits/cpp/deb.out -std=c++26 -Werror -Wall -Wctad-maybe-unsupported -Wextra -Wshadow -Wundef -Wunused-template -Wno-unused-command-line-argument -Wno-attributes -Wno-pessimizing-move -Wno-noexcept-type -Wno-atomic-alignment -Wno-reserved-module-identifier -Wdeprecated-copy -Wdeprecated-copy-dtor -Wno-user-defined-literals -Wno-tautological-compare -Wsign-compare -Wunused-variable -Wunused-parameter -Wunreachable-code -Wno-unused-local-typedef -Wno-local-type-template-args -Wno-c++11-extensions -Wno-unknown-pragmas -Wno-pass-failed -Wno-mismatched-new-delete -Wno-redundant-move -Wno-self-move -D_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER -D_LIBCPP_ENABLE_EXPERIMENTAL -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_NONE -Werror=thread-safety -Wuser-defined-warnings -g   -lc++ -latomic
> 0 spaits@work-laptop:~/cpp$ 
> ```
> 
> Because of the GCC 13+ prereq. I did not build libc++ for my self, but used the one available on my machine. This way the file has compiled successfully.
> 
> I will still try to run the tests myself.
> 
> Until that if you have any idea what should I, do what wen possibly wrong pleas share that with me.
> 
> Thanks.

It could very well just be a transient error/something that failed and has since been fixed.  See if it still happens after this pending build.

cor3ntin commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -858,6 +859,27 @@ class PackDeductionScope {
+  // Return the size of the saved packs if all of them has the same size.
+  std::optional<unsigned> getSavedPackSizeIfAllEqual() const {
+    if (Packs.size() == 0 ||
+        Packs[0].Saved.getKind() != clang::TemplateArgument::Pack)
```
when is `Packs[0].Saved.getKind() != clang::TemplateArgument::Pack` true?

cor3ntin replied:
for that matter when is `Packs.size() != 0` ?

spaits replied:
The reason I added them is because I wanted to be careful. What if someone constructs a `PackDeductionScope` with a non pack `TemplateArgument`.

Should I replace them with insertions?

cor3ntin replied:
I think you can remove it entirely, there is an assert in `.pack_size();` - that should simplify that code a bit !

spaits replied:
Removed this precondition. I have also removed a redundant condition from the checking of the equivalence of pack sizes.

spaits committed:
Add new test case

spaits committed:
Add test cases and remove redundant conditions

cor3ntin reviewed (approved):
Thanks for working on this!
The changes looks good to me.
Will you need me to merge them on your behalf?

spaits commented:
Thank you for reviewing.

Before merge we should take a look at the CI.

It still fails for libc++ 26 suite, with the same reason as before.

I wanted to reproduce the issue. I built libc++ with the clang++ I compiled:
```
cmake -S "../runtimes" -GNinja \
          -DLLVM_ENABLE_RUNTIMES="libcxx;libcxxabi;libunwind" \
          -DLIBCXX_CXX_ABI=libcxxabi \
          -DCMAKE_BUILD_TYPE=Debug \
          -DLIBCXX_TEST_PARAMS="std=c++26" \
          -DLIBCXXABI_TEST_PARAMS="std=c++26" \
          -DCMAKE_CXX_COMPILER=/home/spaits/repo/llvm-project/build/bin/clang++ \
          -DLLVM_DIR=/home/spaits/repo/llvm-project/build/lib/cmake/llvm
```
Then I run the test:
```
 /home/spaits/repo/llvm-project/build/bin/llvm-lit -a /home/spaits/repo/llvm-project/libcxx/test/libcxx/gdb/gdb_pretty_printer_test.sh.cpp
```

Everything worked fine for me.
I don't know why the CI always fails for my PR with the same reason. It looks like an env issue in the CI, but it can not be that since other PRs are getting green runs.

spaits commented:
@cor3ntin I see that you have CI runs that fail with exactly the same reason as my runs: https://buildkite.com/llvm-project/clang-ci/builds/10874#018d49c2-1224-4939-9430-0e5a2be796a9 .

cor3ntin commented:
It's unrelated to your changes. Configuration issue on the CI, probably

spaits commented:
> It's unrelated to your changes. Configuration issue on the CI, probably

Okay. Then I will merge this.

Thank you very much for reviewing my PR and helping me.
